physical/postgresql.go

```diff
@@ -175,3 +199,143 @@ func (m *PostgreSQLBackend) List(prefix string) ([]string, error) {
 
 	return keys, nil
 }
+
+func (m *PostgreSQLBackend) HAEnabled() bool {
+	return true
+}
+
+// PostgreSQLLock implements the Lock interface for PostgreSQL
+type PostgreSQLLock struct {
+	client *sql.DB
+	table  string
+	leader chan struct{}
+
+	key     string
+	value   string
+	vaultID string
+}
+
+// LockWith initializes a Postgres backend lock
+func (m *PostgreSQLBackend) LockWith(key, value string) (Lock, error) {
+	id, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, err
+	}
+	return &PostgreSQLLock{
+		client:  m.client,
+		table:   m.lock_table,
+		leader:  make(chan struct{}),
+		key:     key,
+		value:   value,
+		vaultID: id,
+	}, nil
+}
+
+// Lock grabs a lock, or waits until it is available
+func (m *PostgreSQLLock) Lock(stopCh <-chan struct{}) (<-chan struct{}, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-stopCh
+		cancel()
+	}()
+	ticker := time.NewTicker(PostgreSQLLockRetryInterval)
+	defer ticker.Stop()
+	for {
+		_, err := m.client.ExecContext(
+			ctx,
+			"DELETE FROM "+m.table+" WHERE expiration < now()",
+		)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
The locking protocol could race if two servers start at the same time.  Also, the `key` isn't very sophisticated so I wouldn't recommend using it for much of anything right now (it's important to make it part of the lock, but I wouldn't rely on it exclusively):

https://github.com/hashicorp/vault/blob/f75f5b0addf00cd0ea16ff0c03f8ae93398ae9e5/vault/core.go#L39-L41
https://github.com/hashicorp/vault/blob/f75f5b0addf00cd0ea16ff0c03f8ae93398ae9e5/vault/core.go#L736-L737
https://github.com/hashicorp/vault/blob/f75f5b0addf00cd0ea16ff0c03f8ae93398ae9e5/vault/core.go#L1445

I think it's necessary to do something like the following:

1. I'd change the table definition to be something like the following ([range types FTW](https://www.postgresql.org/docs/current/static/rangetypes.html)):

```sql
-- Optional:
--
-- CREATE EXTENSION IF NOT EXISTS btree_gist;
--
-- if you want to incorporate the `key` into the lock domain - good practice, not required.  Not all installations have the `btree_gist` extension.

CREATE TABLE vault_lock (
  vault_id   TEXT COLLATE "C" NOT NULL,
  key        TEXT COLLATE "C",
  value      TEXT COLLATE "C",

  -- lock_lease
  lock_lease tstzrange NOT NULL,

  -- row_expiration is the time that we should unconditionally remove this row.  row_expiration should be at least 2x lock_ttl.
  row_expiration TIMESTAMP WITHOUT TIME ZONE NOT NULL DEFAULT NOW() + '30 minutes'::INTERVAL,
  EXCLUDE USING gist (/* key WITH = ,*/ lock_lease WITH &&)
);

CREATE INDEX vault_id_idx ON vault_lock(vault_id);
-- CREATE INDEX vault_lock_row_expiration_idx ON vault_lock(row_expiration); -- only suggested if the row_expiration is very large and a large number of rows accumulate.  This is not necessary but couldn't hurt much.
CREATE INDEX vault_key_idx ON vault_lock(vault_id) WHERE key IS NULL;
```

2. The easy part is is cleaning old rows: `DELETE FROM vault_lock WHERE row_expiration < NOW()`.  Simple and easy to understand.  This could be run by any node (ideally just the one that has the lock _after_ it acquires or renews its lock).

3. Registration of a node is a prerequisite: `INSERT INTO vault_lock (key, value, vault_id) VALUES ($1, $2, $3)`

4. Lock acquisition is an `UPDATE`: `UPDATE vault_lock SET key = $1, lock_lease = ('\'[' || NOW() || ',' || NOW() + $2 || ']\'')::tstzrange WHERE vault_id = $3 AND key IS NULL`;

5. If the above lock has been acquired, you can expire all old rows (query from step 2) `DELETE`.

6. Lock renewal is the following `UPDATE`: `UPDATE vault_lock SET lock_lease = (\'[' || lower(lock_lease) || ',' || NOW() + $1 || ']\'')::tstzrange WHERE vault_id = $2 AND key = $3`

For followers attempting to acquire the lock, poll using:

```sql
SELECT TRUE AS lock_held, vault_id AS lock_held_by, upper(lock_lease) - NOW() AS remaining_lock_duration FROM vault_lock WHERE lock_lease @> NOW()
```

If zero rows come back, attempt to acquire the lock.  If the lock acquisition fails, there was a race and someone else acquired the lock and the follower should go back to polling.  It's tempting to use the `remaining_lock_duration` as the time that the followers should sleep but that also means that the lock won't be acquired until the the lease expires.  Maybe do that if the `poll_interval` is set to `0s`.  ?  Food for thought.